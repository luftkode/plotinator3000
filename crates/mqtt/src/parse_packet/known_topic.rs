use egui_plot::PlotPoint;
use serde::Deserialize;
use strum_macros::{Display, EnumString};

use crate::{
    data::listener::{MqttData, MqttTopicData},
    util,
};

/// Known topics can have custom payloads that have an associated known packet structure
/// which allows recognizing and parsing them appropriately
#[derive(EnumString, Display)]
pub(crate) enum KnownTopic {
    #[strum(serialize = "debug/sensors/temperature")]
    DebugSensorsTemperature,
    #[strum(serialize = "debug/sensors/humidity")]
    DebugSensorsHumidity,
    #[strum(serialize = "debug/sensors/pressure")]
    DebugSensorsPressure,
    #[strum(serialize = "debug/sensors/mag")]
    DebugSensorsMag,
    #[strum(serialize = "debug/sensors/gps")]
    DebugSensorsGps,
    #[strum(serialize = "speed")]
    PilotDisplaySpeed,
    #[strum(serialize = "altitude")]
    PilotDisplayAltitude,
    #[strum(serialize = "heading")]
    PilotDisplayHeading,
    #[strum(serialize = "closest_line")]
    PilotDisplayClosestLine,
}

/// Debug packet with a single value
/// e.g. { "value": 70.56164 }
#[derive(Deserialize)]
pub(crate) struct DebugSensorPacket {
    value: f64,
}

/// A pair of values and timestamp
/// e.g. {"value": 3167.38561, "timestamp": "1742662585.527223099"}
#[derive(Deserialize)]
pub(crate) struct ValueWithTimestampString {
    value: f64,
    timestamp: String, // Format determined by `data +%s.%N`
}

/// Debug packet with multiple values
/// e.g. { "value1": 44.50188, "value2": 41.58077 }
#[derive(Deserialize)]
pub(crate) struct DebugSensorsGps {
    value1: f64,
    value2: f64,
}

impl KnownTopic {
    pub(crate) fn parse_packet(self, p: &str) -> Result<MqttData, serde_json::Error> {
        match self {
            Self::DebugSensorsTemperature
            | Self::DebugSensorsHumidity
            | Self::DebugSensorsPressure => {
                let sp: DebugSensorPacket = serde_json::from_str(p)?;
                Ok(self.into_single_mqtt_data(sp.value))
            }
            Self::DebugSensorsGps => {
                let sp: DebugSensorsGps = serde_json::from_str(p)?;
                let td1 = MqttTopicData::single(self.subtopic_str("lat"), sp.value1);
                let td2 = MqttTopicData::single(self.subtopic_str("lon"), sp.value2);
                let d = MqttData::multiple(vec![td1, td2]);
                Ok(d)
            }
            Self::DebugSensorsMag => {
                let values: Vec<ValueWithTimestampString> =
                    serde_json::from_str(p).expect("Debug failure");
                let mut points: Vec<PlotPoint> = vec![];
                for v in values {
                    let t = util::parse_timestamp_to_nanos_f64(&v.timestamp)
                        .expect("failed parsing timestamp");
                    let p = PlotPoint::new(t, v.value);
                    points.push(p);
                }
                let td = MqttTopicData::multiple(self.to_string(), points);
                Ok(MqttData::single(td))
            }
            Self::PilotDisplaySpeed => {
                let p = serde_json::from_str::<PilotDisplaySpeedPacket>(p)?;
                let value = p
                    .speed
                    .parse()
                    .expect("Failed to parse PilotDisplaySpeedPacket");
                Ok(self.into_single_mqtt_data(value))
            }
            Self::PilotDisplayAltitude => {
                let p: PilotDisplayAltitudePacket = serde_json::from_str(p)?;
                let value = p
                    .height
                    .parse()
                    .expect("Failed to parse PilotDisplayAltitudePacket");
                Ok(self.into_single_mqtt_data(value))
            }
            Self::PilotDisplayHeading => {
                let p: PilotDisplayHeadingPacket = serde_json::from_str(p)?;
                let value = p
                    .heading
                    .parse()
                    .expect("Failed to parse PilotDisplayHeadingPacket");
                Ok(self.into_single_mqtt_data(value))
            }
            Self::PilotDisplayClosestLine => {
                let p: PilotDisplayClosestLinePacket = serde_json::from_str(p)?;
                Ok(self.into_single_mqtt_data(p.distance))
            }
        }
    }

    // Converts a value to a simple MqttData with just a single topic and point
    // this is appropriate for very simple topics with just a single point per message
    fn into_single_mqtt_data(self, value: f64) -> MqttData {
        let topic_data = MqttTopicData::single(self.to_string(), value);
        MqttData::single(topic_data)
    }

    // Returns a string with the topic name appended with [subtopic_name]
    //
    // helper function to construct topic strings with a 'subvalue' specifier
    // for topics that essentially have subtopics, meaning they receive payload
    //  with multiple different kind of values, e.g. a gps
    // that publishes both longitude and latitude
    fn subtopic_str(&self, subtopic_name: &str) -> String {
        let mut topic_with_subvalue = self.to_string();
        // yes this could be a single line of format!("[{subtopic_name}]")
        // but this potentially has better performance
        topic_with_subvalue.push('[');
        topic_with_subvalue.push_str(subtopic_name);
        topic_with_subvalue.push(']');
        topic_with_subvalue
    }
}

#[derive(Deserialize)]
pub struct PilotDisplaySpeedPacket {
    #[serde(rename(deserialize = "Speed"))]
    speed: String,
}

#[derive(Deserialize)]
pub struct PilotDisplayAltitudePacket {
    #[serde(rename(deserialize = "Height"))]
    height: String,
}
#[derive(Deserialize)]
pub struct PilotDisplayHeadingPacket {
    #[serde(rename(deserialize = "Heading"))]
    heading: String,
}

#[derive(Deserialize)]
pub struct PilotDisplayClosestLinePacket {
    distance: f64,
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;
    use serde_json::json;

    #[test]
    fn test_enum_parse_packet() {
        let s = "debug/sensors/temperature".to_owned();
        let e = KnownTopic::from_str(&s).unwrap();
        let payload = r#"{ "value": 40 }"#;
        let p = e.parse_packet(payload).unwrap();
        dbg!(p);
    }

    #[test]
    fn test_parse_pilot_display_speed_packet() {
        let s = "speed".to_owned();
        let e = KnownTopic::from_str(&s).unwrap();
        let payload = json!({
            "Speed": "12.433",
        })
        .to_string();
        let p = e.parse_packet(&payload).unwrap();
        dbg!(p);
    }

    #[test]
    fn test_parse_pilot_display_closest_line() {
        let t = "closest_line";
        let known_topic = KnownTopic::from_str(t).unwrap();
        let payload = r#"{"id": 12, "flight_line": "L501100", "distance": 1.84167211, "mode": "automatic", "filename": "20231023_Bremervoerde_Combined_300_NS_32N.kml"}"#;
        let p = known_topic.parse_packet(payload).unwrap();
        dbg!(p);
    }

    #[test]
    fn test_parse_debug_buffered_packet() {
        let topic = KnownTopic::DebugSensorsMag;
        let payload = r#"[{"value": 4125.26202, "timestamp": "1742661659.597146009"},{"value": 5319.64538, "timestamp": "1742661659.597977050"},{"value": 3088.24687, "timestamp": "1742661659.598809170"},{"value": 3032.34963, "timestamp": "1742661659.599677984"},{"value": 3220.23746, "timestamp": "1742661659.600710924"}]"#;
        let _decoded: Vec<ValueWithTimestampString> = serde_json::from_str(payload).unwrap();
        let known_topic = KnownTopic::from_str(&topic.to_string()).unwrap();
        let data = known_topic.parse_packet(payload).unwrap();
        dbg!(data);
    }
}
