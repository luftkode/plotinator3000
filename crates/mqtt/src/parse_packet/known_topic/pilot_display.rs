use std::num::ParseFloatError;

use serde::Deserialize;

#[derive(Deserialize)]
pub struct PilotDisplaySpeedPacket {
    #[serde(rename(deserialize = "Speed"))]
    speed: String,
}

impl PilotDisplaySpeedPacket {
    pub fn speed(self) -> Result<f64, ParseFloatError> {
        self.speed.parse()
    }
}

#[derive(Deserialize)]
pub struct PilotDisplayAltitudePacket {
    #[serde(rename(deserialize = "Height"))]
    height: String,
}

impl PilotDisplayAltitudePacket {
    pub fn height(self) -> Result<f64, ParseFloatError> {
        self.height.parse()
    }
}

#[derive(Deserialize)]
pub struct PilotDisplayHeadingPacket {
    #[serde(rename(deserialize = "Heading"))]
    heading: String,
}

impl PilotDisplayHeadingPacket {
    pub fn heading(self) -> Result<f64, ParseFloatError> {
        self.heading.parse()
    }
}

#[derive(Deserialize)]
pub struct PilotDisplayClosestLinePacket {
    distance: f64,
}

impl PilotDisplayClosestLinePacket {
    pub fn distance(&self) -> f64 {
        self.distance
    }
}
