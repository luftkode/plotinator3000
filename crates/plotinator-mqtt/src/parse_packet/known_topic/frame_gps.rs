use chrono::{DateTime, Utc};
use plotinator_log_if::prelude::{GeoAltitude, GeoPoint};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub(crate) struct FrameGpsPacket {
    pub(crate) gps_nr: u8,
    pub(crate) timestamp: u64,
    /// Fix type. 0 is no fix, 3 is 3D fix
    pub(crate) mode: u8,
    pub(crate) gps_status: FrameGpsStatus,
    pub(crate) position: FrameGpsPosition,
    pub(crate) speed: Option<f32>,
    pub(crate) gps_time: DateTime<Utc>,
}

impl FrameGpsPacket {
    /// Try to extract a [`GeoPoint`] from the [`FrameGpsPacket`] if it has at least longitude and latitude
    #[inline]
    pub(crate) fn maybe_get_geopoint(&self) -> Option<GeoPoint> {
        let ts = self.gps_time.timestamp_nanos_opt()? as f64;
        let lat = self.position.lat?;
        let lon = self.position.lon?;
        let mut geo = GeoPoint::new(ts, (lat.into(), lon.into()));
        if let Some(alt) = self.position.alt {
            let geo_alt = GeoAltitude::Gnss(alt.into());
            geo = geo.with_altitude(geo_alt);
        }
        if let Some(spd) = self.speed {
            geo = geo.with_speed(spd.into());
        }

        Some(geo)
    }
}

#[derive(Debug, Deserialize)]
pub(crate) struct FrameGpsPosition {
    pub(crate) lat: Option<f32>,
    pub(crate) lon: Option<f32>,
    pub(crate) alt: Option<f32>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct FrameGpsStatus {
    pub(crate) hdop: Option<f32>,
    pub(crate) pdop: Option<f32>,
    pub(crate) vdop: Option<f32>,
    pub(crate) satellites: u8,
}

#[cfg(test)]
mod tests {
    use std::str::FromStr as _;

    use crate::parse_packet::known_topic::KnownTopic;

    use super::*;
    use plotinator_test_util::*;

    #[test]
    fn test_parse_packet() -> TestResult {
        let payload = r#"{"gps_nr":2,"timestamp":1757321363140914802,"mode":3,"gps_status":{"hdop":0.8,"vdop":1.5,"pdop":1.7,"satellites":10},"gps_time":"2025-09-08T08:49:23.000Z","position":{"lat":56.217778333,"lon":10.147778333,"alt":66.62},"speed":0.052}"#;
        let decoded_json: FrameGpsPacket = serde_json::from_str(payload)?;
        let known_topic = KnownTopic::from_str(&KnownTopic::TcFrameGps.to_string())?;
        let mqtt_data = known_topic.parse_packet(payload)?.unwrap();

        insta::assert_debug_snapshot!(decoded_json);
        insta::assert_debug_snapshot!(mqtt_data);
        Ok(())
    }
}
