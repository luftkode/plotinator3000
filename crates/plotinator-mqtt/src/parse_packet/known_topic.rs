use anyhow::bail;
use chrono::{TimeZone as _, Utc};
use egui_plot::PlotPoint;
use pilot_display::{PilotDisplayCoordinates, PilotDisplayRemainingDistance};
use plotinator_log_if::prelude::GeoPoint;
use serde::Deserialize;
use strum_macros::{Display, EnumString};

use crate::{
    data::listener::{MqttData, MqttTopicData},
    parse_packet::known_topic::frame_gps::FrameGpsPacket,
    util,
};

pub(crate) mod frame_gps;
pub(crate) mod pilot_display;

/// Known topics can have custom payloads that have an associated known packet structure
/// which allows recognizing and parsing them appropriately
#[derive(EnumString, Display)]
pub(crate) enum KnownTopic {
    /// Timestamped packets from `frame-gps`
    #[strum(serialize = "dt/tc/frame-gps/x/gps")]
    TcFrameGps,
    #[strum(serialize = "dt/njord/frame-gps/x/gps")]
    NjordFrameGps,
    #[strum(serialize = "dt/blackbird/sky-ubx/x/coordinates")]
    PilotDisplayCoordinates,
    #[strum(serialize = "dt/blackbird/pd-backend/remaining-distance")]
    PilotDisplayRemainingDistance,
    #[strum(serialize = "$SYS/broker/uptime")]
    SYSBrokerUptime,
    // We cannot meaningfully plot this, but we use it to show the version when choosing a broker to connect to
    #[strum(serialize = "$SYS/broker/version")]
    SYSBrokerVersion,
    #[strum(serialize = "debug/sensors/temperature")]
    DebugSensorsTemperature,
    #[strum(serialize = "debug/sensors/humidity")]
    DebugSensorsHumidity,
    #[strum(serialize = "debug/sensors/pressure")]
    DebugSensorsPressure,
    #[strum(serialize = "debug/sensors/mag")]
    DebugSensorsMag,
    #[strum(serialize = "debug/sensors/gps")]
    DebugSensorsGps,
}

/// Debug packet with a single value
/// e.g. { "value": 70.56164 }
#[derive(Deserialize)]
pub(crate) struct DebugSensorPacket {
    value: f64,
}

/// A pair of values and timestamp
/// e.g. {"value": 3167.38561, "timestamp": "1742662585.527223099"}
#[derive(Deserialize)]
pub(crate) struct ValueWithTimestampString {
    value: f64,
    timestamp: String, // Format determined by `data +%s.%N`
}

/// Debug packet with multiple values
/// e.g. { "value1": 44.50188, "value2": 41.58077 }
#[derive(Deserialize)]
pub(crate) struct DebugSensorsGps {
    value1: f64,
    value2: f64,
}

impl KnownTopic {
    #[allow(
        clippy::too_many_lines,
        reason = "This kind of match statement will just be long when each topic needs a branch"
    )]
    #[inline]
    pub(crate) fn parse_packet(self, p: &str) -> anyhow::Result<Option<MqttData>> {
        match self {
            Self::TcFrameGps | Self::NjordFrameGps => {
                let p: FrameGpsPacket = match serde_json::from_str(p) {
                    Ok(p) => p,
                    Err(e) => {
                        bail!("Failed deserialising frame-gps packet: {e} - Packet: {p}");
                    }
                };
                let timestamp = p.timestamp as f64;

                let system_time = Utc.timestamp_nanos(p.timestamp as i64);
                let offset = system_time
                    .signed_duration_since(p.gps_time)
                    .num_milliseconds() as f64;

                let id = p.gps_nr;

                let mut topic_data: Vec<MqttTopicData> = Vec::with_capacity(11);

                if let Some(lat) = p.position.lat {
                    topic_data.push(MqttTopicData::single_with_ts(
                        self.subtopic_str(&format!("GP{id} lat")),
                        lat.into(),
                        timestamp,
                    ));
                }
                if let Some(lon) = p.position.lon {
                    topic_data.push(MqttTopicData::single_with_ts(
                        self.subtopic_str(&format!("GP{id} lon")),
                        lon.into(),
                        timestamp,
                    ));
                }
                if let Some(alt) = p.position.alt {
                    topic_data.push(MqttTopicData::single_with_ts(
                        self.subtopic_str(&format!("GP{id} alt")),
                        alt.into(),
                        timestamp,
                    ));
                };

                topic_data.push(MqttTopicData::single_with_ts(
                    self.subtopic_str(&format!("GP{id} mode")),
                    p.mode as f64,
                    timestamp,
                ));

                if let Some(hdop) = p.gps_status.hdop {
                    topic_data.push(MqttTopicData::single_with_ts(
                        self.subtopic_str(&format!("GP{id} hdop")),
                        hdop.into(),
                        timestamp,
                    ));
                }
                if let Some(vdop) = p.gps_status.vdop {
                    topic_data.push(MqttTopicData::single_with_ts(
                        self.subtopic_str(&format!("GP{id} vdop")),
                        vdop.into(),
                        timestamp,
                    ));
                }
                if let Some(pdop) = p.gps_status.pdop {
                    topic_data.push(MqttTopicData::single_with_ts(
                        self.subtopic_str(&format!("GP{id} pdop")),
                        pdop.into(),
                        timestamp,
                    ));
                }
                topic_data.push(MqttTopicData::single_with_ts(
                    self.subtopic_str(&format!("GP{id} satellites")),
                    p.gps_status.satellites.into(),
                    timestamp,
                ));

                if let Some(speed) = p.speed {
                    topic_data.push(MqttTopicData::single_with_ts(
                        self.subtopic_str(&format!("GP{id} speed")),
                        speed.into(),
                        timestamp,
                    ));
                }
                topic_data.push(MqttTopicData::single_with_ts(
                    self.subtopic_str(&format!("GP{id} offset ms")),
                    offset,
                    timestamp,
                ));
                if let Some(geo_point) = p.maybe_get_geopoint() {
                    topic_data.push(MqttTopicData::single_geopoint(
                        self.subtopic_str(&id.to_string()),
                        geo_point,
                    ));
                }
                Ok(Some(MqttData::multiple(topic_data)))
            }
            Self::PilotDisplayCoordinates => {
                let p: PilotDisplayCoordinates = serde_json::from_str(p)?;

                let lat = MqttTopicData::single(self.subtopic_str("lat"), p.lat());
                let lon = MqttTopicData::single(self.subtopic_str("lon"), p.lon());

                let geo_point = GeoPoint::new(util::now_timestamp(), (p.lat(), p.lon()));
                let geo_data = MqttTopicData::single_geopoint(self.to_string(), geo_point);

                let data = MqttData::multiple(vec![lat, lon, geo_data]);
                Ok(Some(data))
            }
            Self::PilotDisplayRemainingDistance => {
                let p: PilotDisplayRemainingDistance = serde_json::from_str(p)?;
                Ok(Some(self.into_single_mqtt_data(p.distance())))
            }
            // Debug topics for development and for inspiration for how to implement parsing of various kinds of
            // topics and payloads
            Self::DebugSensorsTemperature
            | Self::DebugSensorsHumidity
            | Self::DebugSensorsPressure => {
                let sp: DebugSensorPacket = serde_json::from_str(p)?;
                Ok(Some(self.into_single_mqtt_data(sp.value)))
            }
            Self::DebugSensorsGps => {
                let sp: DebugSensorsGps = serde_json::from_str(p)?;
                let td1 = MqttTopicData::single(self.subtopic_str("lat"), sp.value1);
                let td2 = MqttTopicData::single(self.subtopic_str("lon"), sp.value2);
                let d = MqttData::multiple(vec![td1, td2]);
                Ok(Some(d))
            }
            Self::DebugSensorsMag => {
                let values: Vec<ValueWithTimestampString> =
                    serde_json::from_str(p).expect("Debug failure");
                let mut points: Vec<PlotPoint> = vec![];
                for v in values {
                    let t = util::parse_timestamp_to_nanos_f64(&v.timestamp)
                        .expect("failed parsing timestamp");
                    let p = PlotPoint::new(t, v.value);
                    points.push(p);
                }
                let td = MqttTopicData::multiple(self.to_string(), points);
                Ok(Some(MqttData::single(td)))
            }
            Self::SYSBrokerUptime => {
                // Example payload: '2256144 seconds'
                let uptime_str = p.trim_end_matches(" seconds");
                let uptime: f64 = uptime_str.parse()?;
                Ok(Some(self.into_single_mqtt_data(uptime)))
            }
            Self::SYSBrokerVersion => {
                // Does not make sense to plot
                Ok(None)
            }
        }
    }

    // Converts a value to a simple MqttData with just a single topic and point
    // this is appropriate for very simple topics with just a single point per message
    #[inline]
    fn into_single_mqtt_data(self, value: f64) -> MqttData {
        let topic_data = MqttTopicData::single(self.to_string(), value);
        MqttData::single(topic_data)
    }

    // Returns a string with the topic name appended with [subtopic_name]
    //
    // helper function to construct topic strings with a 'subvalue' specifier
    // for topics that essentially have subtopics, meaning they receive payload
    //  with multiple different kind of values, e.g. a gps
    // that publishes both longitude and latitude
    #[inline]
    fn subtopic_str(&self, subtopic_name: &str) -> String {
        let mut topic_with_subvalue = self.to_string();
        // yes this could be a single line of format!("[{subtopic_name}]")
        // but this potentially has better performance
        topic_with_subvalue.push('[');
        topic_with_subvalue.push_str(subtopic_name);
        topic_with_subvalue.push(']');
        topic_with_subvalue
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr as _;

    use super::*;
    use plotinator_test_util::*;
    use serde_json::json;

    #[test]
    fn test_enum_parse_packet() -> TestResult {
        let known_topic = KnownTopic::from_str("debug/sensors/temperature")?;
        let payload = r#"{ "value": 40 }"#;
        let mqtt_data = known_topic.parse_packet(payload)?;
        dbg!(&mqtt_data);
        assert!(mqtt_data.is_some());
        Ok(())
    }

    #[test]
    fn test_parse_pilot_display_coordinate_packet() -> TestResult {
        let known_topic = KnownTopic::from_str("dt/blackbird/sky-ubx/x/coordinates")?;
        let payload = json!({
            "lon": 10.1473,"lat": 56.2179
        })
        .to_string();
        let mqtt_data = known_topic.parse_packet(&payload)?;
        dbg!(&mqtt_data);
        assert!(mqtt_data.is_some());
        Ok(())
    }

    #[test]
    fn test_parse_pilot_display_remaining_distance_packet() -> TestResult {
        let known_topic = KnownTopic::from_str("dt/blackbird/pd-backend/remaining-distance")?;
        let payload = json!({
            "distance": 1560.514601457434
        })
        .to_string();
        let mqtt_data = known_topic.parse_packet(&payload)?;
        dbg!(&mqtt_data);
        assert!(mqtt_data.is_some());
        Ok(())
    }

    #[test]
    fn test_parse_debug_buffered_packet() -> TestResult {
        let topic = KnownTopic::DebugSensorsMag;
        let payload = r#"[{"value": 4125.26202, "timestamp": "1742661659.597146009"},{"value": 5319.64538, "timestamp": "1742661659.597977050"},{"value": 3088.24687, "timestamp": "1742661659.598809170"},{"value": 3032.34963, "timestamp": "1742661659.599677984"},{"value": 3220.23746, "timestamp": "1742661659.600710924"}]"#;
        let _decoded: Vec<ValueWithTimestampString> = serde_json::from_str(payload)?;
        let known_topic = KnownTopic::from_str(&topic.to_string())?;
        let mqtt_data = known_topic.parse_packet(payload)?;
        dbg!(mqtt_data);
        Ok(())
    }
}
