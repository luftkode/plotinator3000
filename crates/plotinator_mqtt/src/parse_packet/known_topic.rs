use egui_plot::PlotPoint;
use pilot_display::{
    PilotDisplayAltitudePacket, PilotDisplayClosestLinePacket, PilotDisplayHeadingPacket,
    PilotDisplaySpeedPacket,
};
use serde::Deserialize;
use strum_macros::{Display, EnumString};

use crate::{
    data::listener::{MqttData, MqttTopicData},
    util,
};

pub(crate) mod pilot_display;

/// Known topics can have custom payloads that have an associated known packet structure
/// which allows recognizing and parsing them appropriately
#[derive(EnumString, Display)]
pub(crate) enum KnownTopic {
    #[strum(serialize = "debug/sensors/temperature")]
    DebugSensorsTemperature,
    #[strum(serialize = "debug/sensors/humidity")]
    DebugSensorsHumidity,
    #[strum(serialize = "debug/sensors/pressure")]
    DebugSensorsPressure,
    #[strum(serialize = "debug/sensors/mag")]
    DebugSensorsMag,
    #[strum(serialize = "debug/sensors/gps")]
    DebugSensorsGps,
    #[strum(serialize = "speed")]
    PilotDisplaySpeed,
    #[strum(serialize = "altitude")]
    PilotDisplayAltitude,
    #[strum(serialize = "heading")]
    PilotDisplayHeading,
    #[strum(serialize = "closest_line")]
    PilotDisplayClosestLine,
    #[strum(serialize = "$SYS/broker/uptime")]
    SYSBrokerUptime,
    // We cannot meaningfully plot this, but we use it to show the version when choosing a broker to connect to
    #[strum(serialize = "$SYS/broker/version")]
    SYSBrokerVersion
}

/// Debug packet with a single value
/// e.g. { "value": 70.56164 }
#[derive(Deserialize)]
pub(crate) struct DebugSensorPacket {
    value: f64,
}

/// A pair of values and timestamp
/// e.g. {"value": 3167.38561, "timestamp": "1742662585.527223099"}
#[derive(Deserialize)]
pub(crate) struct ValueWithTimestampString {
    value: f64,
    timestamp: String, // Format determined by `data +%s.%N`
}

/// Debug packet with multiple values
/// e.g. { "value1": 44.50188, "value2": 41.58077 }
#[derive(Deserialize)]
pub(crate) struct DebugSensorsGps {
    value1: f64,
    value2: f64,
}

impl KnownTopic {
    pub(crate) fn parse_packet(self, p: &str) -> anyhow::Result<Option<MqttData>> {
        match self {
            Self::PilotDisplaySpeed => {
                let p = serde_json::from_str::<PilotDisplaySpeedPacket>(p)?;
                let value = p.speed()?;
                Ok(Some(self.into_single_mqtt_data(value)))
            }
            Self::PilotDisplayAltitude => {
                let p: PilotDisplayAltitudePacket = serde_json::from_str(p)?;
                let value = p.height()?;
                Ok(Some(self.into_single_mqtt_data(value)))
            }
            Self::PilotDisplayHeading => {
                let p: PilotDisplayHeadingPacket = serde_json::from_str(p)?;
                let value = p.heading()?;
                Ok(Some(self.into_single_mqtt_data(value)))
            }
            Self::PilotDisplayClosestLine => {
                let p: PilotDisplayClosestLinePacket = serde_json::from_str(p)?;
                let distance = p.distance();
                Ok(Some(self.into_single_mqtt_data(distance)))
            }
            // Debug topics for development and for inspiration for how to implement parsing of various kinds of
            // topics and payloads
            Self::DebugSensorsTemperature
            | Self::DebugSensorsHumidity
            | Self::DebugSensorsPressure => {
                let sp: DebugSensorPacket = serde_json::from_str(p)?;
                Ok(Some(self.into_single_mqtt_data(sp.value)))
            }
            Self::DebugSensorsGps => {
                let sp: DebugSensorsGps = serde_json::from_str(p)?;
                let td1 = MqttTopicData::single(self.subtopic_str("lat"), sp.value1);
                let td2 = MqttTopicData::single(self.subtopic_str("lon"), sp.value2);
                let d = MqttData::multiple(vec![td1, td2]);
                Ok(Some(d))
            }
            Self::DebugSensorsMag => {
                let values: Vec<ValueWithTimestampString> =
                    serde_json::from_str(p).expect("Debug failure");
                let mut points: Vec<PlotPoint> = vec![];
                for v in values {
                    let t = util::parse_timestamp_to_nanos_f64(&v.timestamp)
                        .expect("failed parsing timestamp");
                    let p = PlotPoint::new(t, v.value);
                    points.push(p);
                }
                let td = MqttTopicData::multiple(self.to_string(), points);
                Ok(Some(MqttData::single(td)))
            }
            Self::SYSBrokerUptime => {
                // Example payload: '2256144 seconds'
                let uptime_str = p.trim_end_matches(" seconds");
                let uptime: f64 = uptime_str.parse()?;
                Ok(Some(self.into_single_mqtt_data(uptime)))
            }
            Self::SYSBrokerVersion => {
                // Does not make sense to plot
                Ok(None)
            }
        }
    }

    // Converts a value to a simple MqttData with just a single topic and point
    // this is appropriate for very simple topics with just a single point per message
    fn into_single_mqtt_data(self, value: f64) -> MqttData {
        let topic_data = MqttTopicData::single(self.to_string(), value);
        MqttData::single(topic_data)
    }

    // Returns a string with the topic name appended with [subtopic_name]
    //
    // helper function to construct topic strings with a 'subvalue' specifier
    // for topics that essentially have subtopics, meaning they receive payload
    //  with multiple different kind of values, e.g. a gps
    // that publishes both longitude and latitude
    fn subtopic_str(&self, subtopic_name: &str) -> String {
        let mut topic_with_subvalue = self.to_string();
        // yes this could be a single line of format!("[{subtopic_name}]")
        // but this potentially has better performance
        topic_with_subvalue.push('[');
        topic_with_subvalue.push_str(subtopic_name);
        topic_with_subvalue.push(']');
        topic_with_subvalue
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;
    use serde_json::json;
    use test_util::*;

    #[test]
    fn test_enum_parse_packet() -> TestResult {
        let known_topic = KnownTopic::from_str("debug/sensors/temperature")?;
        let payload = r#"{ "value": 40 }"#;
        let mqtt_data = known_topic.parse_packet(payload)?;
        dbg!(mqtt_data);
        Ok(())
    }

    #[test]
    fn test_parse_pilot_display_speed_packet() -> TestResult {
        let known_topic = KnownTopic::from_str("speed")?;
        let payload = json!({
            "Speed": "12.433",
        })
        .to_string();
        let mqtt_data = known_topic.parse_packet(&payload)?;
        dbg!(mqtt_data);
        Ok(())
    }

    #[test]
    fn test_parse_pilot_display_closest_line() -> TestResult {
        let known_topic = KnownTopic::from_str("closest_line")?;
        let payload = r#"{"id": 12, "flight_line": "L501100", "distance": 1.84167211, "mode": "automatic", "filename": "20231023_Bremervoerde_Combined_300_NS_32N.kml"}"#;
        let mqtt_data = known_topic.parse_packet(payload)?;
        dbg!(mqtt_data);
        Ok(())
    }

    #[test]
    fn test_parse_debug_buffered_packet() -> TestResult {
        let topic = KnownTopic::DebugSensorsMag;
        let payload = r#"[{"value": 4125.26202, "timestamp": "1742661659.597146009"},{"value": 5319.64538, "timestamp": "1742661659.597977050"},{"value": 3088.24687, "timestamp": "1742661659.598809170"},{"value": 3032.34963, "timestamp": "1742661659.599677984"},{"value": 3220.23746, "timestamp": "1742661659.600710924"}]"#;
        let _decoded: Vec<ValueWithTimestampString> = serde_json::from_str(payload)?;
        let known_topic = KnownTopic::from_str(&topic.to_string())?;
        let mqtt_data = known_topic.parse_packet(payload)?;
        dbg!(mqtt_data);
        Ok(())
    }
}
