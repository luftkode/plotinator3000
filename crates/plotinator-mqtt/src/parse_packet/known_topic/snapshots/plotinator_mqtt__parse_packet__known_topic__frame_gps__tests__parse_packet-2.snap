---
source: crates/plotinator-mqtt/src/parse_packet/known_topic/frame_gps.rs
expression: mqtt_data
---
MqttData {
    inner: Topics(
        [
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 lat]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 56.217777252197266,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 lon]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 10.147778511047363,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 alt]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 66.62000274658203,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 mode]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 3.0,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 hdop]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 0.800000011920929,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 vdop]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 1.5,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 pdop]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 1.7000000476837158,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 satellites]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 10.0,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 speed]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 0.052000001072883606,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[GP2 offset ms]",
                payload: Point(
                    PlotPoint {
                        x: 1.7573213631409147e18,
                        y: 140.0,
                    },
                ),
            },
            MqttTopicData {
                topic: "dt/tc/frame-gps/x/gps[2]",
                payload: GeoPoint(
                    GeoPoint {
                        timestamp: 1.757321363e18,
                        position: POINT(10.147778511047363 56.217777252197266),
                        heading: None,
                        altitude: Some(
                            Gnss(
                                66.62000274658203,
                            ),
                        ),
                        speed: Some(
                            0.052000001072883606,
                        ),
                    },
                ),
            },
        ],
    ),
}
